source/common/compressor/zlib_compressor_impl.h

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include "envoy/compressor/compressor.h"
+
+#include "common/common/assert.h"
+
+#include "zlib.h"
+
+namespace Envoy {
+namespace Compressor {
+
+/**
+ * Deflate implementation of compressor's interface.
+ */
+class ZlibCompressorImpl : public Compressor {
+public:
+  ZlibCompressorImpl();
+  ~ZlibCompressorImpl();
+
+  enum CompressionLevel {
+    default_compression = Z_DEFAULT_COMPRESSION,
+    best = Z_BEST_COMPRESSION,
+    speed = Z_BEST_SPEED,
+  };
+
+  enum CompressionStrategy {
+    filtered = Z_FILTERED,
+    huffman = Z_HUFFMAN_ONLY,
+    rle = Z_RLE,
+    fixed = Z_FIXED,
+    default_strategy = Z_DEFAULT_STRATEGY,
+  };
+
+  /**
+   * Gets current total bytes passed into the compressor
+   * @return current bytes moved into the compressor.
+   */
+  uint64_t getTotalIn();
+
+  /**
+   * Gets current total bytes moved out of the compressor
+   * @return bytes returned by the compressor.
+   */
+  uint64_t getTotalOut();
+
+  /**
+   * Sets how much memory should be allocated by zlib
+   * for the internal compression state.
+   * @param mem_level level of allocated memory, default = 8 | max = 9.
+   */
+  void setMemoryLevel(uint mem_level);
+
+  /**
+   * Buffer size for feeding data to and pulling data from the zlib routines.
+   * For this implementation, it should never be less than the length of evbuffer
+   * passsed into ZlibCompressorImpl::start(Buffer::Instance& in, Buffer::Instance& out).
+   * @param chunk available memory to deflate data, default = 4096.
+   */
+  void setChunk(uint64_t chunk);
+
+  /**
+   * The windowBits parameter shall be a base 2 logarithm of the window size
+   * to use, and shall be a value between 8 and 15 (15 + 16 to write a simple gzip).
+   * A smaller value will use less memory, but will result in a poorer compression
+   * ratio, while a higher value will give better compression but utilize more memory.
+   * @param window_bits memory window size, default = 31 gzip.
+   */
+  void setWindowBits(int window_bits);
```

Comment:
It looks like these three methods aren't used or tested. Is it necessary to expose these details?

Reply:
I removed all setters except setChunk(), since it might be needed for Gzip filter.